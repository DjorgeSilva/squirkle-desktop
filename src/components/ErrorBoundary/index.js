// @flow

import React from 'react'
import useTranslation from '@atomos/atomizer-utils/hooks/useTranslation'
import styles from './styles.module.css'
import Button from '@atomos/atomix/molecules/Button'

type ErrorOutputPropTypes = {
  titleTranslationKey: string,
  messageTranslationKey: string,
  tryAgainOnClick: () => void
}

function ErrorOutput({
  titleTranslationKey,
  messageTranslationKey,
  tryAgainOnClick
}: ErrorOutputPropTypes): React$Element<*> {
  const { translate } = useTranslation('global')

  return (
    <div className={styles.errorBoundaryContainer}>
      <h1>{translate(titleTranslationKey)}</h1>
      <p>{translate(messageTranslationKey)}</p>
      <div className={styles.tryAgainBtnWrapper}>
        <Button label="Try Again" onClick={tryAgainOnClick} />
      </div>
    </div>
  )
}

type ErrorBoundaryPropTypes = {
  children?: any,
  // For things like error view to override
  hasError?: boolean,
  titleTranslationKey?: string,
  messageTranslationKey?: string
}

type StateType = {
  hasError: boolean
}

class ErrorBoundary extends React.Component<ErrorBoundaryPropTypes, StateType> {
  constructor(props: ErrorBoundaryPropTypes): void {
    super(props)

    this.state = { hasError: false }
  }

  /* eslint-disable handle-callback-err */
  static getDerivedStateFromError(error: Error): { hasError: boolean } {
    // Update state so the next render will show the fallback UI.
    return { hasError: true }
  }

  componentDidCatch(error: Error, errorInfo: any): void {
    // You can also log the error to an error reporting service
    // logErrorToMyService(error, errorInfo);
    console.error(error, errorInfo)
  }

  render(): React$Element<*> | null {
    if (this.state.hasError || this.props.hasError) {
      return (
        <ErrorOutput
          titleTranslationKey={
            this.props.titleTranslationKey || 'error-boundary-default-title'
          }
          messageTranslationKey={
            this.props.messageTranslationKey || 'error-boundary-default-msg'
          }
          tryAgainOnClick={() =>
            this.setState({
              hasError: false
            })
          }
        />
      )
    }

    return this.props.children || null
  }
}

export default ErrorBoundary
