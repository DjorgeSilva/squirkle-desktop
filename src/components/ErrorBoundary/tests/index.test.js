// @flow

import React from 'react'
import { mount } from 'enzyme'
import ErrorBoundary from '../'

jest.mock('@atomos/atomizer-utils/hooks/useTranslation')

describe('ErrorBoundary', (): void => {
  describe('When mounted', (): void => {
    let wrapper

    beforeAll((): void => {
      wrapper = mount(<ErrorBoundary>Some child</ErrorBoundary>)
    })

    it('renders correctly', (): void => {
      expect(wrapper).toMatchSnapshot()
    })

    describe('then it errors', (): void => {
      beforeAll((): void => {
        wrapper.setState({ hasError: true })
      })

      it('renders correctly', (): void => {
        expect(wrapper).toMatchSnapshot()
      })
    })
  })
})
