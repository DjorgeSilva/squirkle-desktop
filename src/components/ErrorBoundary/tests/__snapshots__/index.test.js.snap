// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary When mounted renders correctly 1`] = `"Some child"`;

exports[`ErrorBoundary When mounted then it errors renders correctly 1`] = `
<div
  className="errorBoundaryContainer"
>
  <h1>
    error-boundary-default-title
  </h1>
  <p>
    error-boundary-default-msg
  </p>
  <div
    className="tryAgainBtnWrapper"
  >
    <div
      className="buttonContainer default undefined"
    >
      <div
        className="button md enabled action"
        onClick={[Function]}
        title="Try Again"
      >
        <div
          className="contentContainer left"
        >
          <div
            className="labelWrapper"
          >
            <div
              className="labelContainer default undefined labelContainer wrap"
            >
              <div
                className="label md enabled left unselectable none"
                title="Try Again"
              >
                Try Again
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
